htmresearch/algorithms/extended_temporal_memory.py

```diff
@@ -21,538 +21,1134 @@
-                                                       connections):
-          connections.createSynapse(segment,
-                                    presynapticCell,
-                                    self.initialPermanence)
+    matchingSegments = (
+      connections.segmentForFlatIdx(i)
+      for i in xrange(len(numActivePotential))
+      if numActivePotential[i] >= minThreshold
+    )
 
+    maxSegmentsPerCell = connections.maxSegmentsPerCell
+    segmentKey = lambda segment: (segment.cell * maxSegmentsPerCell
+                                  + segment.idx)
 
-  def bestMatchingCell(self, cells, activeCells, activeApicalCells, connections, apicalConnections):
+    if learn:
+      for segment in activeSegments:
+        connections.recordSegmentActivity(segment)
+      connections.startNewIteration()
+
+    return (sorted(activeSegments, key = segmentKey),
+            sorted(matchingSegments, key = segmentKey),
+            numActiveConnected,
+            numActivePotential)
+
+
+  @staticmethod
+  def getLeastUsedCell(random, cells, connections):
+    """ Gets the cell with the smallest number of segments.
+    Break ties randomly.
+    @param random (Object)
+    Random number generator. Gets mutated.
+    @param cells (list)
+    Indices of cells.
+    @param connections (Object)
+    Connections instance for the TM.
+    @return (int) Cell index.
+    """
+    leastUsedCells = []
+    minNumSegments = float("inf")
+    for cell in cells:
+      numSegments = connections.numSegments(cell)
+
+      if numSegments < minNumSegments:
+        minNumSegments = numSegments
+        leastUsedCells = []
+
+      if numSegments == minNumSegments:
+        leastUsedCells.append(cell)
+
+    i = random.getUInt32(len(leastUsedCells))
+    return leastUsedCells[i]
+
+
+  @staticmethod
+  def growSynapses(connections, random, segment, numDesiredNewSynapes,
+                   internalCandidates, externalCandidates, initialPermanence):
+    """ Creates nDesiredNewSynapes synapses on the segment passed in if
+    possible, choosing random cells from the previous winner cells that are
+    not already on the segment.
+    @param  connections        (Object) Connections instance for the TM
+    @param random (Object) Random number generator. Gets mutated.
+    @param  segment            (int)    Segment to grow synapses on.
+    @param numDesiredNewSynapes (int)    Desired number of synapses to grow
+    @param internalCandidates    (list)   Winner cells in `t-1`
+    @param externalCandidates (list)  Active cells in `t-1`, external to the layer
+    @param  initialPermanence  (float)  Initial permanence of a new synapse.
+
+    Notes: The process of writing the last value into the index in the array
+    that was most recently changed is to ensure the same results that we get
+    in the c++ implentation using iter_swap with vectors.
+    """
+    candidates = itertools.chain(internalCandidates, externalCandidates)
```

Comment:
For this you'll actually want to create a new list, since the candidates list gets manipulated. The purpose of `itertools.chain` is that it doesn't create a new list, it just creates an iterator that traverses two iterables.